### PR TITLE
Functional logs POC

### PR DESCRIPTION
**Pull request by floriand-e2r**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?**
No

**What kind of change does this PR introduce?** 
Feature

**What is the current behavior?** 
Functional and technical informations are logged on the fly

**What is the new behavior (if this is a feature change)?**
Using the `Reporter` API from powsybl-core functional logs POC ([PR#1666](https://github.com/powsybl/powsybl-core/pull/1666)) for some functional informations to check the concept.

**Does this PR introduce a breaking change or deprecate an API?**
No

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowReportConstants.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * @author Florian Dupuy <florian.dupuy at rte-france.com>
+ */
+public final class OpenLoadFlowReportConstants {
+
+    private OpenLoadFlowReportConstants() {
+    }
+
+    public static final String MISMATCH_TYPED_VALUE = "Mismatch";
+
+    public static final TypedValue INFO_SEVERITY = new TypedValue("OLF_INFO", TypedValue.INFO_LOGLEVEL);
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/OpenLoadFlowReportConstants.java:21]: Why do you need the `OLF` prefix?

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowProvider.java [hunk 2]
```diff
@@ -289,20 +290,29 @@ private LoadFlowResult runDc(Network network, LoadFlowParameters parameters, Ope
...
+    @Override
+    public CompletableFuture<LoadFlowResult> run(Network network, ComputationManager computationManager, String workingVariantId, LoadFlowParameters parameters, Reporter reporter) {
         Objects.requireNonNull(workingVariantId);
         Objects.requireNonNull(parameters);
 
         LOGGER.info("Version: {}", new PowsyblOpenLoadFlowVersion());
 
         OpenLoadFlowParameters parametersExt = getParametersExt(parameters);
 
+        Reporter lfReporter = reporter.createSubReporter("loadFlow", "Load flow on network ${networkId}",
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/OpenLoadFlowProvider.java:305]: What is the aim of the word "loadflow"?

Comment (floriand-e2r) [reply src/main/java/com/powsybl/openloadflow/OpenLoadFlowProvider.java:305]: It is a key identifying the task, this allows to have multilingual support by having several versions (one for each language) of the taskname.

File: src/main/java/com/powsybl/openloadflow/dc/DcLoadFlowEngine.java [hunk 3]
```diff
@@ -139,6 +150,12 @@ public DcLoadFlowResult run() {
             }
         }
 
+        reporter.report(Report.builder()
+            .withKey("loadFlowComplete")
+            .withDefaultMessage("Dc loadflow complete (status=${lfStatus})")
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/dc/DcLoadFlowEngine.java:155]: Typo: `DC`

File: src/main/java/com/powsybl/openloadflow/network/LfNetwork.java [hunk 4]
```diff
@@ -367,8 +377,17 @@ public void logBalance() {
             reactiveLoad += b.getLoadTargetQ() * PerUnit.SB;
         }
 
+        reporter.report(Report.builder()
+            .withKey("networkBalance")
+            .withDefaultMessage("Network ${numNetwork} balance: active generation=${activeGeneration} Mw, active load=${activeLoad} Mw, reactive generation=${reactiveGeneration} MVar, reactive load=${reactiveLoad} MVar")
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/LfNetwork.java:382]: Typo `MW`

File: src/main/java/com/powsybl/openloadflow/ac/DistributedSlackOuterLoop.java [hunk 5]
```diff
@@ -44,13 +47,21 @@ public String getType() {
...
-    public OuterLoopStatus check(OuterLoopContext context) {
+    public OuterLoopStatus check(OuterLoopContext context, Reporter reporter) {
         double slackBusActivePowerMismatch = context.getLastNewtonRaphsonResult().getSlackBusActivePowerMismatch();
         if (Math.abs(slackBusActivePowerMismatch) > SLACK_BUS_P_MAX_MISMATCH) {
 
             ActivePowerDistribution.Result result = activePowerDistribution.run(context.getNetwork(), slackBusActivePowerMismatch);
 
             if (Math.abs(result.getRemainingMismatch()) > ActivePowerDistribution.P_RESIDUE_EPS) {
+                reporter.report(Report.builder()
+                    .withKey("mismatchDistributionFailed")
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/ac/DistributedSlackOuterLoop.java:58]: `mismatchDistributionFailure`?

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowReportConstants.java (hunk 1 above)
Comment (floriand-e2r) [reply src/main/java/com/powsybl/openloadflow/OpenLoadFlowReportConstants.java:21]: It's just to separate the severity between processes, as for instance the specifications for CGMES import might define specific severities, or open load flow users might have others.

File: src/main/java/com/powsybl/openloadflow/dc/DcLoadFlowEngine.java [hunk 6]
```diff
@@ -125,6 +130,12 @@ public DcLoadFlowResult run() {
             status = LoadFlowResult.ComponentResult.Status.CONVERGED;
         } catch (Exception e) {
             status = LoadFlowResult.ComponentResult.Status.FAILED;
+            reporter.report(Report.builder()
+                .withKey("failedLoadFlow")
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/dc/DcLoadFlowEngine.java:134]: `LoadFlowFailure`?

File: src/main/java/com/powsybl/openloadflow/dc/DcLoadFlowEngine.java (hunk 3 above)
Comment (annetill) [reply src/main/java/com/powsybl/openloadflow/dc/DcLoadFlowEngine.java:155]: `loadFlowSuccess` to stay coherent with your previous choice for distribution?

File: src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java [hunk 7]
```diff
@@ -455,10 +457,20 @@ private static LfNetwork create(MutableInt num, List<Bus> buses, List<Switch> sw
         fixDiscreteVoltageControls(lfNetwork, parameters.isMinImpedance());
 
         if (report.generatorsDiscardedFromVoltageControlBecauseNotStarted > 0) {
+            reporter.report(Report.builder()
+                .withKey("generatorsVcDiscarded1")
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java:461]: `notStartedGenerators`?

File: src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java [hunk 8]
```diff
@@ -455,10 +457,20 @@ private static LfNetwork create(MutableInt num, List<Bus> buses, List<Switch> sw
...
+                .withValue("nbGenImpacted", report.generatorsDiscardedFromVoltageControlBecauseNotStarted)
+                .build());
             LOGGER.warn("Network {}: {} generators have been discarded from voltage control because not started",
                     lfNetwork.getNum(), report.generatorsDiscardedFromVoltageControlBecauseNotStarted);
         }
         if (report.generatorsDiscardedFromVoltageControlBecauseMaxReactiveRangeIsTooSmall > 0) {
+            reporter.report(Report.builder()
+                .withKey("generatorsVcDiscarded2")
+                .withDefaultMessage("${nbGenImpacted} generators have been discarded from voltage control because of a too small max reactive range")
+                .withValue("nbGenImpacted", report.generatorsDiscardedFromVoltageControlBecauseMaxReactiveRangeIsTooSmall)
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/network/impl/LfNetworkLoaderImpl.java:472]: `smallReactiveRangeGenerators`?

**Review by annetill — CHANGES_REQUESTED** [review]
Maybe juste some small changes, but it is globally ok!

**Review by annetill — APPROVED** [review]
:-)

**Review by geofjamg — APPROVED** [review]
Ok don't forget to clean the CI et rely on core 4.1.0 release

**Pull request merged**